File: app/features/scanner/core/frame-kernels.ts
/**
 * In-place frame conversions in WebAssembly SIMD (frame-kernels.c, compiled
 * to a wasm32 object; regeneration steps there), bit-identical to OpenCV's
 * cvtColor at about half its cost. The module runs on OpenCV's own memory
 * (cv.wasmMemory, exposed by the opencv-js patch), so Mat data is read and
 * written where it lies. Null where WebAssembly SIMD or the memory is
 * unavailable; callers then use cvtColor.
 */
import { getCV, type Mat } from "./cv";

/** frame-kernels.c compiled with clang --target=wasm32 -O3 -msimd128, base64 */
const FRAME_KERNELS_WASM =
	"AGFzbQEAAAABh4CAgAABYAN/f38AArqAgIAAAgNlbnYPX19saW5lYXJfbWVtb3J5AgAAA2VudhlfX2luZGlyZWN0X2Z1bmN0aW9uX3RhYmxlAXAAAAODgICAAAIAAAeagICAAAIKcmdiYVRvR3JheQAACXJnYmFUb1JnYgABCuSMgIAAAowKBAN/BHsDfwJ7QQAhAwJAIAJBEEkNAEEAIQQgACEFA0AgASAEaiAF/QAAACIG/YkB/QxGJiNLlw4AAEYmI0uXDgAAIgf9ugEiCCAG/YoBIAf9ugEiBv0NBAUGBwwNDg8UFRYXHB0eHyAIIAb9DQABAgMICQoLEBESExgZGhv9rgH9DABAAAAAQAAAAEAAAABAAAAiBv2uAUEP/a0BIAVBEGr9AAAAIgj9iQEgB/26ASIJIAj9igEgB/26ASII/Q0EBQYHDA0ODxQVFhccHR4fIAkgCP0NAAECAwgJCgsQERITGBkaG/2uASAG/a4BQQ/9rQH9hgEgBUEgav0AAAAiCP2JASAH/boBIgkgCP2KASAH/boBIgj9DQQFBgcMDQ4PFBUWFxwdHh8gCSAI/Q0AAQIDCAkKCxAREhMYGRob/a4BIAb9rgFBD/2tASAFQTBq/QAAACII/YkBIAf9ugEiCSAI/YoBIAf9ugEiB/0NBAUGBwwNDg8UFRYXHB0eHyAJIAf9DQABAgMICQoLEBESExgZGhv9rgEgBv2uAUEP/a0B/YYB/Wb9CwAAIAVBwABqIQUgBEEgaiEKIARBEGoiAyEEIAogAk0NAAsLAkAgAiADTQ0AAkAgAiADayILQRBNDQAgACADQQJ0aiEFAkAgASADaiIEIAAgAkECdGpBf2pPDQAgBSABIAJqSQ0BCyACQQ9xIgpBECAKGyIMIANqIAJrIQogAyALIAxraiEDA0AgBCAF/QAAACIIIAf9DQAECAwAAAAAAAAAAAAAAAD9iQH9qQH9DEYmAABGJgAARiYAAEYmAAAiB/21ASAIIAf9DQEFCQ0AAAAAAAAAAAAAAAD9iQH9qQH9DCNLAAAjSwAAI0sAACNLAAAiBv21Af2uASAIIAf9DQIGCg4AAAAAAAAAAAAAAAD9iQH9qQH9DJcOAACXDgAAlw4AAJcOAAAiCP21Af2uAf0MAEAAAABAAAAAQAAAAEAAACIJ/a4BQQ/9rQH9DP8AAAD/AAAA/wAAAP8AAAAiDf1OIAX9AAAQIg4gB/0NAAQIDAAAAAAAAAAAAAAAAP2JAf2pASAH/bUBIA4gB/0NAQUJDQAAAAAAAAAAAAAAAP2JAf2pASAG/bUB/a4BIA4gB/0NAgYKDgAAAAAAAAAAAAAAAP2JAf2pASAI/bUB/a4BIAn9rgFBD/2tASAN/U79hgEgBf0AACAiDiAH/Q0ABAgMAAAAAAAAAAAAAAAA/YkB/akBIAf9tQEgDiAH/Q0BBQkNAAAAAAAAAAAAAAAA/YkB/akBIAb9tQH9rgEgDiAH/Q0CBgoOAAAAAAAAAAAAAAAA/YkB/akBIAj9tQH9rgEgCf2uAUEP/a0BIA39TiAF/QAAMCIOIAf9DQAECAwAAAAAAAAAAAAAAAD9iQH9qQEgB/21ASAOIAf9DQEFCQ0AAAAAAAAAAAAAAAD9iQH9qQEgBv21Af2uASAOIAf9DQIGCg4AAAAAAAAAAAAAAAD9iQH9qQEgCP21Af2uASAJ/a4BQQ/9rQEgDf1O/YYB/Wb9CwAAIAVBwABqIQUgBEEQaiEEIApBEGoiCg0ACwsgASADaiEEIAIgA2shCiAAIANBAnRqIQUDQCAEIAUtAABBxswAbCAFQQFqLQAAQaOWAWxqIAVBAmotAABBlx1sakGAgAFqQQ92OgAAIAVBBGohBSAEQQFqIQQgCkF/aiIKDQALCwvTAgIDfwF7QQYhA0EAIQQCQCACQQZJDQAgACEFIAEhBANAIAQgBf0AAAAiBiAG/Q0AAQIEBQYICQoMDQ4AAAAA/QsAACAFQRBqIQUgBEEMaiEEIANBBGoiAyACTQ0ACyADQXpqIQQLAkAgAiAETQ0AIARBAWohBQJAIAIgBGtBAXFFDQAgASAEQQNsaiIDIAAgBEECdGoiBC0AADoAACADQQFqIAQtAAE6AAAgA0ECaiAELQACOgAAIAUhBAsgAiAFRg0AIAIgBGshAyAAIARBAnRqIQUgASAEQQNsaiEEA0AgBCAFLQAAOgAAIARBAWogBUEBai0AADoAACAEQQJqIAVBAmotAAA6AAAgBEEDaiAFQQRqLQAAOgAAIARBBGogBUEFai0AADoAACAEQQVqIAVBBmotAAA6AAAgBUEIaiEFIARBBmohBCADQX5qIgMNAAsLCwCxgICAAAdsaW5raW5nAgiigICAAAMApAEACnJnYmFUb0dyYXkApAEBCXJnYmFUb1JnYgWQAQAAwICAgAAJcHJvZHVjZXJzAQxwcm9jZXNzZWQtYnkBC0FwcGxlIGNsYW5nGjIxLjAuMCAoY2xhbmctMjEwMC4zLjM0LjIpAJ2BgIAAD3RhcmdldF9mZWF0dXJlcwkrC2J1bGstbWVtb3J5Kw9idWxrLW1lbW9yeS1vcHQrFmNhbGwtaW5kaXJlY3Qtb3ZlcmxvbmcrCm11bHRpdmFsdWUrD211dGFibGUtZ2xvYmFscysTbm9udHJhcHBpbmctZnB0b2ludCsPcmVmZXJlbmNlLXR5cGVzKwhzaWduLWV4dCsHc2ltZDEyOA==";

export interface FrameKernels {
	/** cvtColor(src, dst, COLOR_RGBA2GRAY) for a continuous CV_8UC4 `src` */
	rgbaToGray(src: Mat): Mat;
	/** cvtColor(src, dst, COLOR_RGBA2RGB) for a continuous CV_8UC4 `src` */
	rgbaToRgb(src: Mat): Mat;
}

let loaded: FrameKernels | null | undefined;

/** The kernels, instantiated on first use; null when this runtime cannot run them. */
export function frameKernels(): FrameKernels | null {
	if (loaded !== undefined) return loaded;
	const cv = getCV();
	const memory = (cv as unknown as { wasmMemory?: WebAssembly.Memory })
		.wasmMemory;
	try {
		if (!memory) throw new Error("OpenCV memory not exposed");
		const { exports } = new WebAssembly.Instance(
			new WebAssembly.Module(
				Uint8Array.from(atob(FRAME_KERNELS_WASM), (c) => c.charCodeAt(0)),
			),
			{
				env: {
					__linear_memory: memory,
					__indirect_function_table: new WebAssembly.Table({
						initial: 0,
						element: "anyfunc",
					}),
				},
			},
		);
		const convert =
			(name: string, type: number) =>
			(src: Mat): Mat => {
				const dst = new cv.Mat(src.rows, src.cols, type);
				(exports[name] as (src: number, dst: number, n: number) => void)(
					(src.data as Uint8Array).byteOffset,
					(dst.data as Uint8Array).byteOffset,
					src.rows * src.cols,
				);
				return dst;
			};
		loaded = {
			rgbaToGray: convert("rgbaToGray", cv.CV_8UC1),
			rgbaToRgb: convert("rgbaToRgb", cv.CV_8UC3),
		};
	} catch {
		loaded = null;
	}
	return loaded;
}

File: app/features/scanner/core/frame-kernels.c
/*
 * In-place color conversions for core/frame-kernels.ts, bit-identical to
 * OpenCV's 8-bit cvtColor: RGBA2GRAY is its 15-bit fixed point
 * (R·9798 + G·19235 + B·3735 + 2^14) >> 15, RGBA2RGB drops alpha. The
 * module imports OpenCV's own memory, so it reads and writes Mat data where
 * it lies.
 *
 * Regenerate FRAME_KERNELS_WASM in frame-kernels.ts after an edit (Apple
 * clang ships the wasm32 target; no linker is needed):
 *   clang --target=wasm32 -O3 -msimd128 -nostdlib -c frame-kernels.c -o frame-kernels.o
 *   base64 -i frame-kernels.o | tr -d '\n'
 * The object is instantiated as is: it imports only its memory (and an unused
 * table), so keep the functions free of stack use, data and calls.
 */
#include <wasm_simd128.h>

typedef unsigned char u8;
typedef unsigned int u32;

static inline v128_t grayOf4(const u8* p, v128_t coef, v128_t half) {
  v128_t px = wasm_v128_load(p);
  v128_t lo = wasm_i32x4_dot_i16x8(wasm_u16x8_extend_low_u8x16(px), coef);
  v128_t hi = wasm_i32x4_dot_i16x8(wasm_u16x8_extend_high_u8x16(px), coef);
  v128_t sum = wasm_i32x4_add(wasm_i32x4_shuffle(lo, hi, 0, 2, 4, 6),
                              wasm_i32x4_shuffle(lo, hi, 1, 3, 5, 7));
  return wasm_u32x4_shr(wasm_i32x4_add(sum, half), 15);
}

__attribute__((export_name("rgbaToGray")))
void rgbaToGray(const u8* src, u8* dst, u32 n) {
  const v128_t coef = wasm_i16x8_make(9798, 19235, 3735, 0, 9798, 19235, 3735, 0);
  const v128_t half = wasm_i32x4_splat(1 << 14);
  u32 i = 0;
  for (; i + 16 <= n; i += 16) {
    const u8* p = src + i * 4;
    v128_t w01 = wasm_u16x8_narrow_i32x4(grayOf4(p, coef, half), grayOf4(p + 16, coef, half));
    v128_t w23 = wasm_u16x8_narrow_i32x4(grayOf4(p + 32, coef, half), grayOf4(p + 48, coef, half));
    wasm_v128_store(dst + i, wasm_u8x16_narrow_i16x8(w01, w23));
  }
  for (; i < n; i++) {
    const u8* p = src + i * 4;
    dst[i] = (u8)((p[0] * 9798 + p[1] * 19235 + p[2] * 3735 + (1 << 14)) >> 15);
  }
}

__attribute__((export_name("rgbaToRgb")))
void rgbaToRgb(const u8* src, u8* dst, u32 n) {
  u32 i = 0;
  // 16-byte stores of 12 valid bytes; the next group overwrites the rest, so
  // stop while a whole store still fits inside dst
  for (; i + 6 <= n; i += 4) {
    v128_t px = wasm_v128_load(src + i * 4);
    wasm_v128_store(dst + i * 3, wasm_i8x16_shuffle(px, px, 0, 1, 2, 4, 5, 6, 8, 9, 10, 12, 13, 14, 0, 0, 0, 0));
  }
  for (; i < n; i++) {
    dst[i * 3] = src[i * 4];
    dst[i * 3 + 1] = src[i * 4 + 1];
    dst[i * 3 + 2] = src[i * 4 + 2];
  }
}
